File: src/main/media/library.ts
/**
 * 图库的磁盘布局与路径安全。
 *
 * <libraryRoot>/
 *   index.db              SQLite 索引
 *   originals/<一级>/<二级>/<文件名>   原图（按站点分类归档，方便脱离本应用也能翻）
 *   thumbs/<id>.jpg      缩略图
 *   .tmp/                下载中转，成功后 rename
 */
import { mkdir, rm, stat } from 'node:fs/promises'
import { existsSync } from 'node:fs'
import { extname, join, normalize, relative, resolve, sep } from 'node:path'

export class Library {
  readonly root: string
  readonly originalsDir: string
  readonly thumbsDir: string
  readonly tmpDir: string
  readonly dbPath: string

  constructor(root: string) {
    this.root = resolve(root)
    this.originalsDir = join(this.root, 'originals')
    this.thumbsDir = join(this.root, 'thumbs')
    this.tmpDir = join(this.root, '.tmp')
    this.dbPath = join(this.root, 'index.db')
  }

  async ensure(): Promise<void> {
    await mkdir(this.originalsDir, { recursive: true })
    await mkdir(this.thumbsDir, { recursive: true })
    await mkdir(this.tmpDir, { recursive: true })
  }

  /** 相对路径 -> 绝对路径（并阻止越权访问） */
  resolveInside(relPath: string): string {
    const abs = resolve(this.root, relPath)
    const rel = relative(this.root, abs)
    if (rel.startsWith('..') || (rel !== '' && rel.split(sep)[0] === '..') || resolve(abs) === resolve(this.root)) {
      throw new Error(`路径越界: ${relPath}`)
    }
    return abs
  }

  exists(relPath: string): boolean {
    try {
      return existsSync(this.resolveInside(relPath))
    } catch {
      return false
    }
  }

  async size(relPath: string): Promise<number | null> {
    try {
      const s = await stat(this.resolveInside(relPath))
      return s.size
    } catch {
      return null
    }
  }

  /** 为某个条目计算原图的相对路径 */
  originalRelPath(input: {
    id: number
    plate: string | null
    word: string | null
    slug: string
    ext: string
    naming: 'id-slug' | 'id' | 'pixiv' | 'hash'
    pixivId?: string | null
    sha256?: string | null
  }): string {
    const plate = safeSegment(input.plate || '未分类')
    const word = safeSegment(input.word || '全部')
    const ext = input.ext.replace(/^\./, '').toLowerCase() || 'jpg'
    const slug = safeSegment(input.slug).slice(0, 48)
    let name: string
    switch (input.naming) {
      case 'id':
        name = `${input.id}.${ext}`
        break
      case 'pixiv':
        name = input.pixivId ? `pid${input.pixivId}.${ext}` : `${input.id}.${ext}`
        break
      case 'hash':
        name = `${(input.sha256 ?? String(input.id)).slice(0, 16)}.${ext}`
        break
      case 'id-slug':
      default:
        name = slug && slug !== String(input.id) ? `${input.id}_${slug}.${ext}` : `${input.id}.${ext}`
        break
    }
    return toPosix(join('originals', plate, word, name))
  }

  thumbRelPath(id: number): string {
    return toPosix(join('thumbs', `${id}.jpg`))
  }

  tmpPath(name: string): string {
    return join(this.tmpDir, name)
  }

  async remove(relPath: string): Promise<void> {
    await rm(this.resolveInside(relPath), { force: true }).catch(() => {})
  }

  /** 清理下载中转目录里的残留 */
  async cleanupTmp(): Promise<void> {
    await rm(this.tmpDir, { recursive: true, force: true }).catch(() => {})
    await mkdir(this.tmpDir, { recursive: true })
  }
}

/** 去掉文件名里的非法字符，保留中文 */
export function safeSegment(input: string): string {
  return input
    .replace(/[\\/:*?"<>|\u0000-\u001f]/g, '_')
    .replace(/[.\s]+$/g, '')
    .replace(/\s+/g, ' ')
    .trim()
}

/** 由标题生成文件名片段：取标签和 pixiv id 更有辨识度 */
export function buildSlug(parts: (string | null | undefined)[]): string {
  return parts
    .filter((p): p is string => Boolean(p && p.trim()))
    .map((p) => p.trim())
    .join('-')
    .replace(/[\[\]（）()]/g, '')
    .replace(/-+/g, '-')
}

function toPosix(p: string): string {
  return normalize(p).split(sep).join('/')
}

export function extOf(path: string): string {
  return extname(path).replace(/^\./, '').toLowerCase()
}

File: src-tauri/src/library.rs
/**
 * 图库的磁盘布局与路径安全（对应 Electron 版的 src/main/media/library.ts）。
 *
 * <libraryRoot>/
 *   index.db              SQLite 索引
 *   originals/<一级>/<二级>/<文件名>   原图
 *   thumbs/<id>.jpg      缩略图
 *   .tmp/                下载中转
 */
use std::fs;
use std::path::{Path, PathBuf};

#[derive(Clone)]
pub struct Library {
    pub root: PathBuf,
    pub originals_dir: PathBuf,
    pub thumbs_dir: PathBuf,
    pub tmp_dir: PathBuf,
    pub db_path: PathBuf,
}

impl Library {
    pub fn new(root: &str) -> Self {
        let root = PathBuf::from(root);
        Self {
            originals_dir: root.join("originals"),
            thumbs_dir: root.join("thumbs"),
            tmp_dir: root.join(".tmp"),
            db_path: root.join("index.db"),
            root,
        }
    }

    pub fn ensure(&self) -> std::io::Result<()> {
        fs::create_dir_all(&self.originals_dir)?;
        fs::create_dir_all(&self.thumbs_dir)?;
        fs::create_dir_all(&self.tmp_dir)?;
        Ok(())
    }

    /// 相对路径 -> 绝对路径（并阻止越权访问）
    pub fn resolve_inside(&self, rel: &str) -> Option<PathBuf> {
        let rel = rel.replace('\\', "/");
        let joined = self.root.join(rel.replace('/', std::path::MAIN_SEPARATOR_STR));
        let normalized = normalize(&joined);
        let root = normalize(&self.root);
        if normalized == root || !normalized.starts_with(&root) {
            return None;
        }
        Some(normalized)
    }

    pub fn original_rel_path(
        &self,
        id: i64,
        plate: Option<&str>,
        word: Option<&str>,
        slug: &str,
        ext: &str,
        naming: &str,
        pixiv_id: Option<&str>,
        sha256: Option<&str>,
    ) -> String {
        let plate = safe_segment(plate.unwrap_or("未分类"));
        let word = safe_segment(word.unwrap_or("全部"));
        let ext = ext.trim_start_matches('.').to_lowercase();
        let ext = if ext.is_empty() { "jpg".to_string() } else { ext };
        let slug: String = safe_segment(slug).chars().take(48).collect();

        let name = match naming {
            "id" => format!("{id}.{ext}"),
            "pixiv" => match pixiv_id {
                Some(p) if !p.is_empty() => format!("pid{p}.{ext}"),
                _ => format!("{id}.{ext}"),
            },
            "hash" => {
                let h = sha256.unwrap_or("");
                let h = if h.is_empty() { id.to_string() } else { h.to_string() };
                format!("{}.{ext}", &h[..h.len().min(16)])
            }
            _ => {
                if !slug.is_empty() && slug != id.to_string() {
                    format!("{id}_{slug}.{ext}")
                } else {
                    format!("{id}.{ext}")
                }
            }
        };
        format!("originals/{plate}/{word}/{name}")
    }

    pub fn thumb_rel_path(&self, id: i64) -> String {
        format!("thumbs/{id}.jpg")
    }

    pub fn tmp_path(&self, name: &str) -> PathBuf {
        self.tmp_dir.join(name)
    }

    pub fn remove(&self, rel: &str) {
        if let Some(p) = self.resolve_inside(rel) {
            let _ = fs::remove_file(p);
        }
    }

    pub fn cleanup_tmp(&self) {
        let _ = fs::remove_dir_all(&self.tmp_dir);
        let _ = fs::create_dir_all(&self.tmp_dir);
    }
}

fn normalize(p: &Path) -> PathBuf {
    let mut out = PathBuf::new();
    for comp in p.components() {
        match comp {
            std::path::Component::ParentDir => {
                out.pop();
            }
            std::path::Component::CurDir => {}
            other => out.push(other.as_os_str()),
        }
    }
    out
}

/// 去掉文件名里的非法字符，保留中文
pub fn safe_segment(input: &str) -> String {
    let mut s = String::with_capacity(input.len());
    for ch in input.chars() {
        if matches!(ch, '\\' | '/' | ':' | '*' | '?' | '"' | '<' | '>' | '|') || (ch as u32) < 0x20 {
            s.push('_');
        } else {
            s.push(ch);
        }
    }
    let trimmed = s.trim_end_matches(['.', ' ']).to_string();
    trimmed.split_whitespace().collect::<Vec<_>>().join(" ")
}

/// 由标题 / 标签拼出文件名片段
pub fn build_slug(parts: &[Option<&str>]) -> String {
    let joined = parts
        .iter()
        .flatten()
        .map(|p| p.trim())
        .filter(|p| !p.is_empty())
        .collect::<Vec<_>>()
        .join("-");
    let stripped: String = joined
        .chars()
        .filter(|c| !matches!(c, '[' | ']' | '（' | '）' | '(' | ')'))
        .collect();
    let mut out = String::new();
    let mut last_dash = false;
    for ch in stripped.chars() {
        if ch == '-' {
            if last_dash {
                continue;
            }
            last_dash = true;
        } else {
            last_dash = false;
        }
        out.push(ch);
    }
    out
}

pub fn ext_of(path: &str) -> String {
    path.rsplit_once('.')
        .map(|(_, e)| e.to_lowercase())
        .filter(|e| e.len() <= 5 && !e.contains('/'))
        .unwrap_or_default()
}
